File: apps/windows/src/FlashTeX.Editor/MathCaretHighlight.cs
// name: MathCaretHighlight.cs
// purpose: Source -> preview for a caret inside math. Ported from
//   apps/mac/Sources/FlashTeXMac/MathCaretHighlight.swift. A formula is laid out
//   as ordinary `glyph_run` items (math font, script sizes) plus `rule` items
//   (fraction bars, the `\sqrt` overbar), and EVERY cluster and rule of the
//   formula carries one source range: the whole formula including its
//   delimiters (`$...$`, `\[...\]`). No cluster maps its bytes 1:1, so the
//   cluster-matching primitive below returns every glyph cluster of the formula
//   with no exact caret and the rules not at all.
//
//   `CaretHighlights` keeps the exact caret / whole-cluster contract for text
//   and adds the enclosing formula box: when no cluster containing the caret
//   byte maps 1:1 and one source span fans out over several items, the
//   highlight is the union of every item on the page with exactly that span
//   (glyph clusters AND rules). A single non-1:1 cluster keeps the documented
//   whole-cluster fallback.
//
//   The cluster-matching primitive (`CaretMatch`/`Clusters`) is ported from the
//   Swift source's `V2Geometry.clusters(containing:)`
//   (apps/mac/Sources/FlashTeXMac/GlyphRunRenderer.swift), included here because
//   `MathCaretHighlight.swift` depends on it directly and it is the same pure
//   byte-range/geometry family this port's instructions describe.
// author: Claude Sonnet 5
// date: 2026-09-14

using TickRect = FlashTeX.Protocol.RenderingV2.Rect;
using Item = FlashTeX.Protocol.RenderingV2.Item;
using Cluster = FlashTeX.Protocol.RenderingV2.Cluster;
using Caret = FlashTeX.Protocol.RenderingV2.Caret;
using Page = FlashTeX.Protocol.RenderingV2.Page;
using SourceRange = FlashTeX.Protocol.RuntimeV1.SourceRange;

namespace FlashTeX.Editor;

/// <summary>One glyph cluster whose source contains a queried caret byte, with exact caret geometry when derivable.</summary>
public readonly record struct CaretMatch(int ItemIndex, int ClusterIndex, Caret? Caret, IReadOnlyList<TickRect> HitRects);

/// <summary>The enclosing box of one fanned-out source span (a whole formula, or a sub-expression sharing one source range).</summary>
public sealed record FormulaBox(
    SourceRange Source,
    IReadOnlyList<int> ItemIndices,
    int ClusterCount,
    int RuleCount,
    IReadOnlyList<TickRect> Rects,
    TickRect Bounds);

/// <summary>Either an exact/whole-cluster caret, or the enclosing box of a formula the caret is inside.</summary>
public abstract record CaretHighlight
{
    private CaretHighlight()
    {
    }

    /// <summary>Exact caret bar (1:1 cluster with a caret at that byte) or the whole-cluster fallback.</summary>
    public sealed record OfCluster(CaretMatch Match) : CaretHighlight;

    /// <summary>The caret is inside a formula: the enclosing box of every item carrying that span.</summary>
    public sealed record OfFormula(FormulaBox Box) : CaretHighlight;
}

public static class MathCaretHighlight
{
    /// <summary>
    /// Clusters whose sources contain <paramref name="byteOffset"/> of
    /// <paramref name="path"/> (`start &lt;= byte &lt; end`; an empty range
    /// matches only its own offset), with exact caret geometry when derivable.
    /// </summary>
    public static IReadOnlyList<CaretMatch> Clusters(int byteOffset, string path, Page page)
    {
        var output = new List<CaretMatch>();
        for (int index = 0; index < page.Items.Count; index++)
        {
            if (page.Items[index] is not Item.OfGlyphRun { Value: var run })
            {
                continue;
            }
            for (int clusterIndex = 0; clusterIndex < run.Clusters.Count; clusterIndex++)
            {
                var cluster = run.Clusters[clusterIndex];
                var source = FirstSourceContaining(cluster.Sources, path, byteOffset);
                if (source is null)
                {
                    continue;
                }
                Caret? caret = null;
                if (source.EndByte - source.StartByte == cluster.TextEndByte - cluster.TextStartByte)
                {
                    int textByte = cluster.TextStartByte + (byteOffset - source.StartByte);
                    caret = cluster.Carets.FirstOrDefault(c => c.TextByte == textByte);
                }
                output.Add(new CaretMatch(index, clusterIndex, caret, cluster.HitRects));
            }
        }
        return output;
    }

    /// <summary>Every item of <paramref name="page"/> whose sources contain exactly <paramref name="span"/>, as one box.</summary>
    public static FormulaBox? BoxFor(SourceRange span, Page page)
    {
        var indices = new List<int>();
        var rects = new List<TickRect>();
        int clusterCount = 0, ruleCount = 0;
        for (int index = 0; index < page.Items.Count; index++)
        {
            switch (page.Items[index])
            {
                case Item.OfRule { Value: var rule }:
                    if (rule.Sources is null || !rule.Sources.Contains(span))
                    {
                        continue;
                    }
                    ruleCount += 1;
                    indices.Add(index);
                    rects.Add(new TickRect(rule.X, rule.Top, rule.Width, rule.Height));
                    break;
                case Item.OfGlyphRun { Value: var run }:
                    bool member = false;
                    foreach (var cluster in run.Clusters)
                    {
                        if (cluster.Sources is null || !cluster.Sources.Contains(span))
                        {
                            continue;
                        }
                        clusterCount += 1;
                        member = true;
                        rects.AddRange(cluster.HitRects);
                    }
                    if (member)
                    {
                        indices.Add(index);
                    }
                    break;
                default:
                    continue; // images and TikZ paths are never part of a formula box
            }
        }
        if (rects.Count == 0)
        {
            return null;
        }
        long minX = rects[0].X.Value, minTop = rects[0].Top.Value;
        long maxX = rects[0].X.Value + rects[0].Width.Value, maxBottom = rects[0].Top.Value + rects[0].Height.Value;
        for (int i = 1; i < rects.Count; i++)
        {
            var r = rects[i];
            minX = Math.Min(minX, r.X.Value);
            minTop = Math.Min(minTop, r.Top.Value);
            maxX = Math.Max(maxX, r.X.Value + r.Width.Value);
            maxBottom = Math.Max(maxBottom, r.Top.Value + r.Height.Value);
        }
        var bounds = new TickRect(minX, minTop, maxX - minX, maxBottom - minTop);
        return new FormulaBox(span, indices, clusterCount, ruleCount, rects, bounds);
    }

    /// <summary>
    /// Highlights for the caret at <paramref name="byteOffset"/> of <paramref name="path"/> on <paramref name="page"/>:
    /// every 1:1 cluster containing the byte, as <see cref="CaretHighlight.OfCluster"/>; otherwise, for each distinct
    /// fanned-out span containing the byte (carried by two or more items, or by any rule), one
    /// <see cref="CaretHighlight.OfFormula"/> box; a span carried by a single glyph cluster stays
    /// <see cref="CaretHighlight.OfCluster"/> (fallback).
    /// </summary>
    public static IReadOnlyList<CaretHighlight> CaretHighlights(int byteOffset, string path, Page page)
    {
        var matches = Clusters(byteOffset, path, page);
        var exact = new List<CaretHighlight>();
        var fanned = new List<SourceRange>();
        var single = new List<CaretHighlight>();
        foreach (var match in matches)
        {
            if (page.Items[match.ItemIndex] is not Item.OfGlyphRun { Value: var run })
            {
                continue;
            }
            var cluster = run.Clusters[match.ClusterIndex];
            var source = FirstSourceContaining(cluster.Sources, path, byteOffset);
            if (source is null)
            {
                continue;
            }
            if (IsOneToOne(cluster, source))
            {
                exact.Add(new CaretHighlight.OfCluster(match));
            }
            else if (!fanned.Contains(source))
            {
                fanned.Add(source);
                var box = BoxFor(source, page);
                single.Add(box is not null && box.ClusterCount + box.RuleCount > 1
                    ? new CaretHighlight.OfFormula(box)
                    : new CaretHighlight.OfCluster(match));
            }
            // else: another cluster of a span already boxed; covered by the box.
        }
        return exact.Count > 0 ? exact : single;
    }

    /// <summary>Whether <paramref name="cluster"/> maps its source bytes 1:1 onto its text bytes.</summary>
    private static bool IsOneToOne(Cluster cluster, SourceRange source) =>
        source.EndByte - source.StartByte == cluster.TextEndByte - cluster.TextStartByte;

    private static SourceRange? FirstSourceContaining(IReadOnlyList<SourceRange>? sources, string path, int byteOffset)
    {
        if (sources is null)
        {
            return null;
        }
        foreach (var source in sources)
        {
            if (source.Path != path)
            {
                continue;
            }
            bool contains = source.StartByte == source.EndByte
                ? byteOffset == source.StartByte
                : source.StartByte <= byteOffset && byteOffset < source.EndByte;
            if (contains)
            {
                return source;
            }
        }
        return null;
    }
}

File: apps/mac/Sources/FlashTeXMac/MathCaretHighlight.swift
import Foundation
import FlashTeXProtocol

/// Source → preview for a caret inside math (lane mac-navigation-3).
///
/// Measured on flashtex-render 9aaec57a: there is no `math` item kind. A
/// formula is laid out as ordinary `glyph_run` items (math font, script
/// sizes) plus `rule` items (fraction bars, the `\sqrt` overbar), and EVERY
/// cluster and rule of the formula carries one source range — the whole
/// formula including its delimiters (`$…$`, `\[…\]`). No cluster maps its
/// bytes 1:1, so `V2Geometry.clusters(containing:)` returns every glyph
/// cluster of the formula with no exact caret and the rules not at all.
///
/// `caretHighlights` keeps the exact caret / whole-cluster contract for text
/// and adds the enclosing formula box: when no cluster containing the caret
/// byte maps 1:1 and one source span fans out over several items, the
/// highlight is the union of every item on the page with exactly that span
/// (glyph clusters AND rules). A single non-1:1 cluster (`\'e` → é) keeps
/// the documented whole-cluster fallback. If a producer ever maps a math
/// atom 1:1 (`x` of `\frac{x}{y}` → its own bytes), that exact caret wins.
extension V2Geometry {
    /// The enclosing box of one fanned-out source span.
    struct FormulaBox: Equatable {
        /// The span every member item carries (the formula including delimiters).
        var source: RenderingV2.SourceRange
        /// Page item indices in the box, in page order.
        var itemIndices: [Int]
        var clusterCount: Int
        var ruleCount: Int
        /// Every member rectangle (cluster hit rects and rule rects), in ticks.
        var rects: [RenderingV2.Rect]
        /// Union of `rects`, in ticks.
        var bounds: RenderingV2.Rect
    }

    /// The caret's paragraph on a page: every item whose source span lies
    /// within the paragraph's bytes, merged into one rectangle per line row.
    struct ParagraphBand: Equatable {
        /// The paragraph's bytes (blank-line delimited in the source).
        var source: RenderingV2.SourceRange
        /// One rectangle per row the paragraph occupies on this page, in
        /// ticks, top to bottom: the union of the member rectangles that
        /// overlap vertically (a superscript, a fraction bar and its line
        /// share a row).
        var rows: [RenderingV2.Rect]
        /// Member items on this page (glyph runs with at least one member
        /// cluster, plus rules).
        var itemCount: Int
    }

    enum CaretHighlight: Equatable {
        /// Exact caret bar (1:1 cluster with a caret at that byte) or the
        /// whole-cluster fallback, exactly as before.
        case cluster(CaretMatch)
        /// The caret is inside a formula: the enclosing box of every item
        /// carrying that span.
        case formula(FormulaBox)
        /// The caret's whole paragraph, as a faint band per row (only when
        /// the preview follows the caret; always last in the list, so
        /// `first` keeps meaning the caret itself).
        case paragraph(ParagraphBand)
    }

    /// The bytes of the paragraph containing `byte` of `text`: the caret's
    /// line grown over every adjacent non-blank line, up to the newline
    /// before the next blank line. A blank line holds only spaces, tabs or a
    /// CR. A caret on a blank line (or past the end) gets an empty range.
    static func paragraphBounds(containing byte: Int, in text: String) -> Range<Int> {
        var copy = text
        return copy.withUTF8 { b -> Range<Int> in
            let n = b.count
            let byte = max(0, min(byte, n))
            let nl = UInt8(ascii: "\n")
            func isBlank(_ s: Int, _ e: Int) -> Bool {
                var i = s
                while i < e { let c = b[i]; if c != 0x20, c != 0x09, c != 0x0D { return false }; i += 1 }
                return true
            }
            // The caret's line.
            var lineStart = byte
            while lineStart > 0, b[lineStart - 1] != nl { lineStart -= 1 }
            var lineEnd = byte
            while lineEnd < n, b[lineEnd] != nl { lineEnd += 1 }
            guard !isBlank(lineStart, lineEnd) else { return lineStart..<lineStart }
            // Backwards over non-blank lines: `start - 1` is the newline ending the previous line.
            var start = lineStart
            while start > 0 {
                var previous = start - 1
                while previous > 0, b[previous - 1] != nl { previous -= 1 }
                if isBlank(previous, start - 1) { break }
                start = previous
            }
            // Forwards over non-blank lines: `end` is the newline ending the current last line.
            var end = lineEnd
            while end < n {
                var next = end + 1
                while next < n, b[next] != nl { next += 1 }
                if isBlank(end + 1, next) { break }
                end = next
            }
            return start..<end
        }
    }

    /// Every item of `page` whose sources for `path` lie within `paragraph`,
    /// as one band of row rectangles; nil when nothing on the page does.
    static func paragraphBand(for paragraph: Range<Int>, path: String, in page: RenderingV2.Page) -> ParagraphBand? {
        guard !paragraph.isEmpty else { return nil }
        @inline(__always) func inside(_ s: RenderingV2.SourceRange) -> Bool {
            s.path == path && s.startByte >= paragraph.lowerBound && s.endByte <= paragraph.upperBound
                && (s.startByte < s.endByte || paragraph.contains(s.startByte))
        }
        var rects: [RenderingV2.Rect] = []
        var items = 0
        for item in page.items {
            switch item {
            case .rule(let r):
                guard r.sources?.contains(where: inside) == true else { continue }
                items += 1
                rects.append(RenderingV2.Rect(x: r.x, top: r.top, width: r.width, height: r.height))
            case .glyphRun(let run):
                var member = false
                for c in run.clusters where c.sources?.contains(where: inside) == true {
                    member = true
                    rects.append(contentsOf: c.hitRects)
                }
                if member { items += 1 }
            case .image, .path:
                continue // figures are not part of a text band
            }
        }
        guard !rects.isEmpty else { return nil }
        return ParagraphBand(source: RenderingV2.SourceRange(path: path, startByte: paragraph.lowerBound, endByte: paragraph.upperBound),
                             rows: rows(merging: rects), itemCount: items)
    }

    /// Merges `rects` into one rectangle per row: sorted by top, a rectangle
    /// joins the current row when it overlaps it vertically, else opens the
    /// next one. Each row is the union of its members.
    static func rows(merging rects: [RenderingV2.Rect]) -> [RenderingV2.Rect] {
        var rows: [RenderingV2.Rect] = []
        for r in rects.sorted(by: { $0.top < $1.top }) {
            if let last = rows.last, r.top < last.top &+ last.height {
                let x = min(last.x, r.x), top = min(last.top, r.top)
                let right = max(last.x &+ last.width, r.x &+ r.width), bottom = max(last.top &+ last.height, r.top &+ r.height)
                rows[rows.count - 1] = RenderingV2.Rect(x: x, top: top, width: right &- x, height: bottom &- top)
            } else {
                rows.append(r)
            }
        }
        return rows
    }

    /// Whether `match` maps its source bytes 1:1 onto its text bytes (an
    /// exact caret is derivable for some byte of it).
    private static func isOneToOne(_ cluster: RenderingV2.Cluster, _ source: RenderingV2.SourceRange) -> Bool {
        source.endByte - source.startByte == cluster.textEndByte - cluster.textStartByte
    }

    /// Every item of `page` whose sources contain exactly `span`, as one box.
    static func formulaBox(for span: RenderingV2.SourceRange, in page: RenderingV2.Page) -> FormulaBox? {
        var indices: [Int] = [], rects: [RenderingV2.Rect] = []
        var clusters = 0, rules = 0
        for (index, item) in page.items.enumerated() {
            switch item {
            case .rule(let r):
                guard r.sources?.contains(span) == true else { continue }
                rules += 1
                indices.append(index)
                rects.append(RenderingV2.Rect(x: r.x, top: r.top, width: r.width, height: r.height))
            case .glyphRun(let run):
                var member = false
                for c in run.clusters where c.sources?.contains(span) == true {
                    clusters += 1; member = true
                    rects.append(contentsOf: c.hitRects)
                }
                if member { indices.append(index) }
            case .image, .path:
                continue // images and TikZ paths are never part of a formula box
            }
        }
        guard let first = rects.first else { return nil }
        var minX = first.x, minTop = first.top, maxX = first.x &+ first.width, maxBottom = first.top &+ first.height
        for r in rects.dropFirst() {
            minX = min(minX, r.x); minTop = min(minTop, r.top)
            maxX = max(maxX, r.x &+ r.width); maxBottom = max(maxBottom, r.top &+ r.height)
        }
        return FormulaBox(source: span, itemIndices: indices, clusterCount: clusters, ruleCount: rules, rects: rects,
                          bounds: RenderingV2.Rect(x: minX, top: minTop, width: maxX &- minX, height: maxBottom &- minTop))
    }

    /// Highlights for the caret at `byte` of `path` on `page`:
    /// - every 1:1 cluster containing the byte, as `.cluster` (exact caret
    ///   when the producer supplied one at that byte, else its rects);
    /// - otherwise, for each distinct fanned-out span containing the byte
    ///   (carried by two or more items, or by any rule), one `.formula` box;
    /// - a span carried by a single glyph cluster stays `.cluster` (fallback);
    /// - with `paragraph` (the caret's paragraph bytes, `paragraphBounds`),
    ///   one trailing `.paragraph` band when any item of the page lies in it.
    static func caretHighlights(containing byte: Int, path: String, in page: RenderingV2.Page,
                                paragraph: Range<Int>? = nil) -> [CaretHighlight] {
        var out = caretItemHighlights(containing: byte, path: path, in: page)
        if let paragraph, let band = paragraphBand(for: paragraph, path: path, in: page) { out.append(.paragraph(band)) }
        return out
    }

    private static func caretItemHighlights(containing byte: Int, path: String, in page: RenderingV2.Page) -> [CaretHighlight] {
        let matches = clusters(containing: byte, path: path, in: page)
        var exact: [CaretHighlight] = []
        var fanned: [RenderingV2.SourceRange] = []
        var single: [CaretHighlight] = []
        for m in matches {
            guard case .glyphRun(let run) = page.items[m.itemIndex] else { continue }
            let c = run.clusters[m.clusterIndex]
            guard let source = (c.sources ?? []).first(where: { s in
                s.path == path && (s.startByte == s.endByte ? byte == s.startByte : (s.startByte <= byte && byte < s.endByte))
            }) else { continue }
            if isOneToOne(c, source) {
                exact.append(.cluster(m))
            } else if !fanned.contains(source) {
                fanned.append(source)
                if let box = formulaBox(for: source, in: page), box.clusterCount + box.ruleCount > 1 {
                    single.append(.formula(box))
                } else {
                    single.append(.cluster(m))
                }
            } else {
                // Another cluster of a span already boxed: covered by the box.
            }
        }
        return exact.isEmpty ? single : exact
    }
}

extension ShellModel {
    /// What the v2 pane highlights for the editor caret, for the note and
    /// tests: the formula box(es) the caret is inside on the current frame,
    /// with page numbers. Empty when the caret is in no formula.
    func caretFormulaBoxes() -> [(page: Int, box: V2Geometry.FormulaBox)] {
        guard let frame = displayListV2?.frame, let byte = caretByte else { return [] }
        var out: [(page: Int, box: V2Geometry.FormulaBox)] = []
        for page in frame.list.pages {
            for h in V2Geometry.caretHighlights(containing: byte, path: activePath, in: page) {
                if case .formula(let box) = h { out.append((page.number, box)) }
            }
        }
        return out
    }
}
